gdal/ogr/ogrsf_frmts/shape/shpopen.c

```diff
@@ -890,381 +880,159 @@ SHPOpenLLEx( const char * pszLayer, const char * pszAccess, SAHooks *psHooks )
         snprintf( pszMessage, nMessageLen, "Unable to open %s.shp or %s.SHP.",
-                 pszBasename, pszBasename );
+                  pszBasename, pszBasename );
         psHooks->Error( pszMessage );
         free( pszMessage );
-        
-        free( psSHP );
+
         free( pszBasename );
         free( pszFullname );
-        
-        return NULL;
+
+        return( 0 );
     }
-    
+
 /* -------------------------------------------------------------------- */
-/*  Read the file size from the SHP file.                               */
+/*  Read the file size from the SHP file.               */
 /* -------------------------------------------------------------------- */
     pabyBuf = (uchar *) malloc(100);
-    psSHP->sHooks.FRead( pabyBuf, 100, 1, psSHP->fpSHP );
-    
-    psSHP->nFileSize = ((unsigned int)pabyBuf[24] * 256 * 256 * 256
+    psHooks->FRead( pabyBuf, 100, 1, fpSHP );
+
+    nSHPFilesize = ((unsigned int)pabyBuf[24] * 256 * 256 * 256
                         + (unsigned int)pabyBuf[25] * 256 * 256
                         + (unsigned int)pabyBuf[26] * 256
                         + (unsigned int)pabyBuf[27]);
-    if( psSHP->nFileSize < 0xFFFFFFFFU / 2 )
-        psSHP->nFileSize *= 2;
+    if( nSHPFilesize < 0xFFFFFFFFU / 2 )
+        nSHPFilesize *= 2;
     else
-        psSHP->nFileSize = 0xFFFFFFFEU;
-    
-/* -------------------------------------------------------------------- */
-/*  Create the .shx file for existing .shp, if SHP_RESTORE_SHX parameter*/
-/*  was passed.                                                         */
-/* -------------------------------------------------------------------- */
-    if( psSHP->fpSHX == NULL )
+        nSHPFilesize = 0xFFFFFFFEU;
+
+    const char pszSHXAccess[] = "w+b";
+    snprintf( pszFullname, nFullnameLen, "%s.shx", pszBasename );
+    fpSHX = psHooks->FOpen( pszFullname, pszSHXAccess );
+
+    if( fpSHX == NULL )
     {
-        const char pszSHXAccess[] = "w+b";
-        snprintf( pszFullname, nFullnameLen, "%s.shx", pszBasename );
-        psSHP->fpSHX = psHooks->FOpen( pszFullname, pszSHXAccess );
-        
-        if(psSHP->fpSHX == NULL)
+        size_t nMessageLen = strlen( pszBasename ) * 2 + 256;
+        char *pszMessage = (char *) malloc( nMessageLen );
+        snprintf( pszMessage, nMessageLen, "Error opening file %s.shx for writing",
+                 pszBasename );
+        psHooks->Error( pszMessage );
+        free( pszMessage );
+
+        psHooks->FClose( fpSHX );
+
+        free( pszBasename );
+        free( pszFullname );
+
+        return( 0 );
+    }
+
+    psHooks->FSeek( fpSHP, 100, 0 );
+    char *pabySHXHeader = (char *) malloc ( 100 );
+    memcpy( pabySHXHeader, pabyBuf, 100 );
+    psHooks->FWrite( pabySHXHeader, 100, 1, fpSHX );
+
+    unsigned int nCurrentRecordOffset = 0;
+    unsigned int nCurrentSHPOffset = 100;
+    size_t nRealSHXContentSize = 100;
+
+    char *pabyReadedRecord = (char *) malloc ( 8 );
```

Comment:
pabyReadedRecord will be leaked. Should rather be abyReadedRecord[8] since no dynamic allocation is needed